File: typescript/src/graph/dijkstra.ts
// LombokAlgoritma — Dijkstra single-source shortest paths
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { NegativeWeightError } from '../core/errors.js';
import { MinHeap, tupleLess } from '../core/heap.js';
import { type Graph, assertNode, buildAdjList } from './types.js';

/**
 * Shortest distance from `source` to every node (`Infinity` when unreachable).
 * Binary heap keyed by `[distance, node]`, O((V + E) log V). SPEC §6.4.
 *
 * v0.1.x re-sorted the whole frontier on every pop (O(V² log V)).
 *
 * @throws RangeError on a negative edge weight
 */
export function dijkstra(g: Graph, source: number): number[] {
  const adj = buildAdjList(g);
  assertNode(g, source, 'source');
  if (g.edges.some((e) => e.weight < 0)) throw new NegativeWeightError('dijkstra');
  const dist = new Array<number>(g.nodes).fill(Number.POSITIVE_INFINITY);
  dist[source] = 0;
  const heap = new MinHeap<[number, number]>(tupleLess);
  heap.push([0, source]);
  for (let top = heap.pop(); top !== undefined; top = heap.pop()) {
    const [d, u] = top;
    if (d > (dist[u] as number)) continue;
    for (const { to, weight } of adj[u] ?? []) {
      const nd = d + weight;
      if (nd < (dist[to] as number)) {
        dist[to] = nd;
        heap.push([nd, to]);
      }
    }
  }
  return dist;
}

File: python/lombokalgoritma/graph/dijkstra.py
# LombokAlgoritma — Dijkstra single-source shortest paths (SPEC §9.3)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

import heapq
import math

from ..errors import NegativeWeightError
from .types import Graph, assert_node, build_adj_list


def dijkstra(g: Graph, source: int) -> list[float]:
    """Shortest distances from ``source`` (``inf`` when unreachable); heap keyed ``(d, node)``.

    Raises:
        NegativeWeightError: any edge weight is negative.
    """
    adj = build_adj_list(g)
    assert_node(g, source, "source")
    if any(w < 0 for _, _, w in g.edges):
        raise NegativeWeightError("dijkstra")
    dist: list[float] = [math.inf] * len(adj)
    dist[source] = 0
    heap: list[tuple[float, int]] = [(0, source)]
    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]:
            continue
        for v, w in adj[u]:
            nd = d + w
            if nd < dist[v]:
                dist[v] = nd
                heapq.heappush(heap, (nd, v))
    return dist
